src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationKeyComparer.cs

```diff
@@ -29,49 +29,111 @@ public class ConfigurationKeyComparer : IComparer<string>
         /// <returns>Less than 0 if x is less than y, 0 if x is equal to y and greater than 0 if x is greater than y.</returns>
         public int Compare(string? x, string? y)
         {
-            string[] xParts = x?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
-            string[] yParts = y?.Split(_keyDelimiterArray, StringSplitOptions.RemoveEmptyEntries) ?? Array.Empty<string>();
+            ReadOnlySpan<char> xSpan = x.AsSpan();
+            ReadOnlySpan<char> ySpan = y.AsSpan();
+
+            if (x == null || y == null)
+            {
+                return (x is null ? 0 : CountPartsIn(xSpan)) - (y is null ? 0 : CountPartsIn(ySpan));
+            }
```

Comment:
When the string is `null`, then `AsSpan` returns an empty span. So the check for empty span (i.e. string is null) could be moved to `CountPartsIn` then this code becomes:
```suggestion
            ReadOnlySpan<char> xSpan = x.AsSpan();
            ReadOnlySpan<char> ySpan = y.AsSpan();

            if (xSpan.IsEmpty || ySpan.IsEmpty)
            {
                return CountPartsIn(xSpan) - CountPartsIn(ySpan);
            }
```
Saves some branches here.